### ISSUE TITLE
Failing test appear to be re-run though it doesn't interefere with the code of my latest commit

### ISSUE THREAD

**Issue by gturri** [gmu-swe/deflaker#6]
Hi,

I'm experimenting with deflaker, and I have the feeling that it is re-running a failed flaky test instead of using the coverage to reach this conclusion.

What I did is: I created a test branch on a simple repo and did 3 commits:
* A first one which adds DeFlaker to my pom.xml: https://github.com/gturri/aXMLRPC/commit/3701b0569844efaf282f5997d7a126056e34b7c0
* A 2nd one which adds a flaky failure (some tests have 75% chance of failing): https://github.com/gturri/aXMLRPC/commit/cdee0b8488199b6e71e010c5afa3f2d726744209
* A 3rd one which is a dummy change unrelated to the test which should fail: https://github.com/gturri/aXMLRPC/commit/4fa40b958c8630e29141f9afa07ee26cc4dc598b

Steps to reproduce:
* get my test branch:
```
git clone https://github.com/gturri/aXMLRPC
cd aXMLRPC
git checkout test_deflaker 
```

* run the test until there's a test which fails: `mvn clean test`

Expected result:
* the failing test isn't re-run because it uses code unrelated to the latest commit so re-running is pointless to figure out that the test is flaky

Actual result:
There appear to be re-run because I can see in the console output:

```
Failed tests:                                                                                                                           
de.timroes.axmlrpc.serializer.TestSerializers.canSerializeBase64(de.timroes.axmlrpc.serializer.TestSerializers)
  Run 1: TestSerializers.canSerializeBase64:31->assertExpectedSerialized:83 expected:<<base64[>AQIDBA==</base64]>> but was:<<base64[/]>>
  Run 2: TestSerializers.canSerializeBase64:31->assertExpectedSerialized:83 expected:<<base64[>AQIDBA==</base64]>> but was:<<base64[/]>>
  Run 3: TestSerializers.canSerializeBase64:31->assertExpectedSerialized:83 expected:<<base64[>AQIDBA==</base64]>> but was:<<base64[/]>>
  Run 4: TestSerializers.canSerializeBase64:31->assertExpectedSerialized:83 expected:<<base64[>AQIDBA==</base64]>> but was:<<base64[/]>>
  Run 5: TestSerializers.canSerializeBase64:31->assertExpectedSerialized:83 expected:<<base64[>AQIDBA==</base64]>> but was:<<base64[/]>>
[...]
[WARNING] Diff has gone untested: de/timroes/axmlrpc/Call has uncovered changes [18]
[WARNING] FLAKY>> Test de.timroes.axmlrpc.serializer.TestSerializers.canSerializeBase64 was found to be flaky by rerunning it in the same JVM! It failed the first time, then eventually passed.
[WARNING] FLAKY>> Test de.timroes.axmlrpc.serializer.TestSerializers.canSerializeBase64 failed, but did not appear to run any changed code
```

Long story short: Am I using deflaker in a bad way, or am I misunderstanding how it should behave?

**Comment by jon-bell** [comment]
Thanks for your interest in DeFlaker.

Did you run DeFlaker *first* on the prior revision (and observe the test pass), and *then*, in the same directory, git checkout the next revision, then run DeFlaker and observe the test fail? From your description it sounds like DeFlaker may not have a history file.

**Comment by gturri** [comment]
Hi,

thanks for replying so quickly!

Indeed, when I tried yesterday I wasn't aware that DeFlaker needed a history file. 
So I just gave it another try, but I'm still not so sure about what I see:


What I just did this morning:

1. I check out my commit which adds the flaky tests and runs `mvn clean test`. I'm lucky and my test passes.
Not sure if it's relevant but I see in the console output:

```
Diff'ing cdee0b8488199b6e71e010c5afa3f2d726744209 against: 3701b0569844efaf282f5997d7a126056e34b7c0   
de/timroes/base64/Base64 : [Edit [start=4, end=4], Edit [start=96, end=98], Edit [start=167, end=167]]
         Methods to watch:[]                                                                          
```
I'm not really sure why the "Methods to watch" array seems empty: I guess it could have make sense to see my `Base64.encode` method mentioned here since my change is https://github.com/gturri/aXMLRPC/commit/cdee0b8488199b6e71e010c5afa3f2d726744209

2. I now checkout the next commit (which is just a dumb change completely unrelated to the flaky test introduced with the previous commit).
I see that when I previously ran `mvn test` it created in particular the files `.diffCache`, `target/diffcov-tests-rerun` and `target/diffcov.log`, so I guess I have the required history file. I hence go ahead and run `mvn test`. Actually I'm out of luck: all my tests are passing. I hence re-run `mvn test` on this latest commit until a test fails, to observe DeFlaker in action.

Eventually I see that a test fails. But if seems that DeFlaker re-ran this test instead of concluding that it's flaky based on the coverage. Here is the console output which looks relevant to me:

```
C:\misc-repos\aXMLRPC/.git DIFF'ing to HEAD
Diff'ing 4fa40b958c8630e29141f9afa07ee26cc4dc598b against: cdee0b8488199b6e71e010c5afa3f2d726744209
de/timroes/axmlrpc/Call : [Edit [start=18, end=18]]
         Methods to watch:[]
[...]
                                                                                                                                        
-------------------------------------------------------                                                                                 
 T E S T S                                                                                                                              
-------------------------------------------------------                                                                                 
[...]
canSerializeBase64(de.timroes.axmlrpc.serializer.TestSerializers)  Time elapsed: 0.001 sec  <<< FAILURE!                                
org.junit.ComparisonFailure: expected:<<base64[>AQIDBA==</base64]>> but was:<<base64[/]>>                                               
        at org.junit.Assert.assertEquals(Assert.java:115)                                                                               
        at org.junit.Assert.assertEquals(Assert.java:144)                                                                               
        at de.timroes.axmlrpc.serializer.TestSerializers.assertExpectedSerialized(TestSerializers.java:83)                              
        at de.timroes.axmlrpc.serializer.TestSerializers.canSerializeBase64(TestSerializers.java:31)                                    
                                                                                                                                        
canSerializeBase64(de.timroes.axmlrpc.serializer.TestSerializers)  Time elapsed: 0 sec  <<< FAILURE!                                    
org.junit.ComparisonFailure: expected:<<base64[>AQIDBA==</base64]>> but was:<<base64[/]>>                                               
        at org.junit.Assert.assertEquals(Assert.java:115)                                                                               
        at org.junit.Assert.assertEquals(Assert.java:144)                                                                               
        at de.timroes.axmlrpc.serializer.TestSerializers.assertExpectedSerialized(TestSerializers.java:83)                              
        at de.timroes.axmlrpc.serializer.TestSerializers.canSerializeBase64(TestSerializers.java:31)                                    
                                                                                                                                        
Running de.timroes.base64.TestBase64                                                                                                    
Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec - in de.timroes.base64.TestBase64                             
                                                                                                                                        
Results :                                                                                                                               
                                                                                                                                        
Flaked tests:                                                                                                                           
de.timroes.axmlrpc.serializer.TestSerializers.canSerializeBase64(de.timroes.axmlrpc.serializer.TestSerializers)                         
  Run 1: TestSerializers.canSerializeBase64:31->assertExpectedSerialized:83 expected:<<base64[>AQIDBA==</base64]>> but was:<<base64[/]>>
  Run 2: TestSerializers.canSerializeBase64:31->assertExpectedSerialized:83 expected:<<base64[>AQIDBA==</base64]>> but was:<<base64[/]>>
  Run 3: PASS
```

Is there something I'm missing, or something I'm misunderstanding?

**Comment by jon-bell** [comment]
"Methods to watch" is a bit of a misnomer - this is a list of methods that had structural changes that DeFlaker can't track at the statement-level. 

It seems like there's a bug in DeFlaker that's having it still do the reruns here - when I run these, I see the reruns too, but I also get this:
```
[WARNING] Diff has gone untested: de/timroes/axmlrpc/Call has uncovered changes [18]
[WARNING] FLAKY>> Test de.timroes.axmlrpc.serializer.TestSerializers.canSerializeBase64 failed, but did not appear to run any changed code
```
From this output, DeFlaker is declaring it flaky from the coverage criteria.

Unfortunately, as it's setup, DeFlaker will not short-circuit the rerun process. By default, right now, it's configuring things to do reruns regardless of whether the coverage criteria determines it not flaky (you can set `-Ddeflaker.rerunFlakies=0` to get this output without reruns). It will take a bit longer, but I can come back to this if you are interested and hack it up to abort reruns if a test is flaky by coverage.

**Comment by gturri** [comment]
Thanks for those explanations.
